### ISSUE TITLE
Can I call `backward()` several times before calling `optim.step()` with AMP?

### ISSUE THREAD

**Issue by jatentaki** [NVIDIA/apex#558]
I'm sorry for asking here, but I'm not sure which is the preferred location for Q&A regarding apex.

Sometimes when batches that fit in GPU memory are not big enough to keep gradient variance at reasonable level, it is possible to call `loss.backward()` on several consecutive batches (therefore averaging the batch gradients) before executing a single step of the optimizer. My question is whether I can do the same when training with `amp`. The reason why I am not sure is that `amp.scale_loss` takes both the loss and optimizer and I am not exactly sure what are the implications of its operation on the optimizer - are all operations it does on the optimizer "averageable" or will the second run of `backward` incorrectly overwrite some state set in the first run?

**Comment by ptrblck** [comment]
Hi @jatentaki,

yes, gradient accumulation should be possible:
https://nvidia.github.io/apex/advanced.html#gradient-accumulation-across-iterations

Let us know, if you see any issues.

**Comment by shenfe** [comment]
Hi @ptrblck , is there any Apex solution like [torch.nn.parallel.DistributedDataParallel.no_sync](https://pytorch.org/docs/stable/generated/torch.nn.parallel.DistributedDataParallel.html#torch.nn.parallel.DistributedDataParallel.no_sync) when it does not need sync if `iter%iters_to_accumulate != 0` ?

**Comment by shenfe** [comment]
Can I use methods `ddp.enable_allreduce()` and `ddp.disable_allreduce()` of `apex.parallel.DistributedDataParallel`?
What I want is exactly [Skip unnecessary all-reduce if training with DistributedDataParallel and gradient accumulation](https://pytorch.org/tutorials/recipes/recipes/tuning_guide.html#skip-unnecessary-all-reduce-if-training-with-distributeddataparallel-and-gradient-accumulation) but in Apex.
